CASMTRIAGE-7509: Create workaround for USS-1.1 customers (#5667)
Cray-HPE/docs-csm: troubleshooting/known_issues/enable_iscsid_multipathd.md

Reviewer: The prompts need to be pulled out of the copy-pasteable code snippets along with the command output, in doing so the copy-paste buttons will work appropriately in the rendered docs.

The prompts, if important, can be included in the text preceding the code snippet as described [here](https://github.com/Cray-HPE/docs-csm/blob/release/1.7/introduction/documentation_conventions.md#command-prompt-conventions).

Reviewer: There are numerous commands + output examples that need to be broken up still.

# Boot content projection services can fail if `iscsid` and `multipathd` services are not enabled on compute nodes and UANs

## Issue Description

iSCSI based boot content projection which is also known as "Scalable Boot Content Projection" (SBPS) for `rootfs` and `PE` images
is supported in version CSM 1.6.0 and above. On a customer system, using CSM-1.6.0 with USS-1.1.x on compute nodes/ UANs in order to support AARCH64 images, 

> `iscsid` and `multipathd` services are not enabled by default. SBPS fails if these services are not enabled by default on compute/ UANs. 

Reviewer 2: ```suggestion
`iscsid` and `multipathd` services are not enabled by default. SBPS will not be resilient across worker node reboots if these services are not enabled by default on compute/ UANs. 
```

Reviewer 2: Totally OK if you want to keep the existing wording.

## Issue Identification

> Issue can be identified with the below symptoms:

Reviewer 3: Reword: **_This_** issue can be identified **_by the following_** symptoms:

> On compute/ UAN node (iSCSI Initiator) we can observe the following SQUASHFS error messages in the console log:

Reviewer 3: Reword: "On **_a_** compute/UAN node" (and remove the space after the slash)

```text
nid000004:~ # dmesg -T | grep "SQUASHFS error" | head -n  1
[Sat Nov  2 22:32:41 2024] SQUASHFS error: xz decompression failed, data probably corrupt
```

> On compute/ UAN node (iSCSI Initiator) we can observe that the iscsid service is not active:

Reviewer 3: Reword: "On **_a_** compute/UAN node" (and remove the space after the slash)

```bash
nid000004:~ # systemctl status iscsid
● iscsid.service - Open-iSCSI
     Loaded: loaded (/usr/lib/systemd/system/iscsid.service; disabled; preset: disabled)
     Active: active (running) since Wed 2024-11-06 08:16:23 CST; 1 day 4h ago
TriggeredBy: ● iscsid.socket
```

From the `journalctl` logs:

```text
Nov 07 10:24:23 nid000004 iscsid[22286]: iscsid: Kernel reported iSCSI connection 2:0 error (1020 - ISCSI_ERR_TCP_CONN_CLOSE: TCP connection closed) state (3)
Nov 07 10:25:14 nid000004 iscsid[22286]: iscsid: connect to 10.253.0.3:3260 failed (No route to host)
...
Nov 07 10:30:43 nid000004 iscsid[22286]: iscsid: connect to 10.253.0.3:3260 failed (Connection refused)
...
```

## Workaround Description

> Step-1: Get the version of the “csm-packages” from compute node BOS session template

Reviewer 3: All of the "Step-n" need more prominence.
Option 1: Make this a heading by adding `### ` to the beginning of each of the lines.
Option 2: Make it bold with `**Step-1**`,
Then, everywhere where a reference is made to one of the steps, make it bold.

Author: resolved

**Example**

> 

Reviewer 3: Add a comment to explain the example:
"Find the session template name"

Author: Resolved

```bash
ncn-m001:~ # cray bos sessiontemplates  list | grep compute-*
…
enable_cfs = true
name = "compute-25.1.0-alpha2.x86_64-csm-160-rc4”
…
```

```bash
ncn-m001:~ # cray bos sessiontemplates describe compute-25.1.0-alpha2.x86_64-csm-160-rc4  --format json
{…
  "cfs": {
    "configuration": "compute-25.1.0-alpha2-csm-160-rc4"
  },
…
}
```

> 

Reviewer 3: Add a comment to explain the example:
Use the `configuration` value to describe the configuration

Author: Resolved.

```bash
ncn-m001:~ # cray cfs configurations describe compute-25.1.0-alpha2-csm-160-rc4 --format json
{
  "lastUpdated": "2024-11-02T12:15:21Z",
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "d530f0a277c9d5dc9e3cb487d32d6b316757f00e",
      "name": "csm-packages-1.6.0-rc.4", 
      "playbook": "csm_packages.yml"
    }, 
…

> ```

Reviewer 3: Add a final comment:
The `name` identifies product catalog, the version is after `csm-packages-`

Author: Resolved.

Step-2: Get the corresponding csm-config branch (@VCS) from product catalog given “csm-packages-*” version found from Step-1

**Example**

```bash
ncn-m001:~ # kubectl get cm -n services cray-product-catalog -o yaml | yq - r 'data.csm' | grep ^1.6.0-rc.4: -A 10
1.6.0-rc.4:
  configuration:
    clone_url: https://vcs.cmn.fanta.hpc.amslabs.hpecorp.net/vcs/cray/csm-config-management.git
    commit: d530f0a277c9d5dc9e3cb487d32d6b316757f00e
    import_branch: cray/csm/1.27.2

> ```

Reviewer 3: Add a comment:
The `import_branch` will be used below.

Author: Resolved.

Step-3: Log into VCS and clone “csm-config-management.git” @ VCS

```bash
ncn-m001:/home/ # USERNAME=$( kubectl get secrets -n services vcs-user-credentials -o json | jq -r .data.vcs_username | base64 -d  )
ncn-m001:/home/ # PSWD=$( kubectl get secrets -n services vcs-user-credentials -o json | jq -r .data.vcs_password | base64 -d  )
ncn-m001:/home/ # git clone https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git
```

**Note:** use above $USERNAME and $PSWD for VCS login

> Step-4: Apply fix against import_branch found in Step-2

Reviewer 3: add back-quotes around `import_branch` to make it stand out.

**Example**

```bash
ncn-m001:/home/ # cd csm-config-management
ncn-m001:/home/csm-config-management # git checkout cray/csm/1.27.2
ncn-m001:/home/csm-config-management # git checkout -b CASMTRIAGE-7509
```
**Note: ** `cray/csm/1.27.2` is a target branch and `CASMTRIAGE-7509` is a new branch

Add new role to enable `iscsid` and `multipathd` service:

```bash

> ncn-m001:~/csm-config-management # cat roles/csm.enable_iscsid_multipathd/tasks/main.yml

Reviewer 3: Turn this into a script they can cut-n-paste.  Make this first line:
```
cat > roles/csm.enable_iscsid_multipathd/tasks/main.yml << EOF
```

---
- name: Ensure iscsid service is started
  ansible.builtin.systemd:
    name: iscsid
    state: started
    enabled: true

- name: Ensure multipathd service is started
  ansible.builtin.systemd:
        name: multipathd
        state: started
        enabled: true

> ncn-m001:~/csm-config-management #

Reviewer 3: And then change this line to just
```
EOF
```

```

> Apply below changes to “csm-config-management/csm_packages.yml” to Application-nodes only play and Compute-nodes only play under “csm_services” in order to enable `iscsid` and `multipathd` services.

Reviewer 3: Change to: "Apply **_the following_** changes"
Also use back-tics, not double quotes, for `csm-config-management/csm_packages.yml`

```bash
diff --git a/csm_packages.yml b/csm_packages.yml
index e3366f8..b223aec 100755
--- a/csm_packages.yml
+++ b/csm_packages.yml
@@ -137,6 +137,9 @@
       vars:
         packages: "{{application_csm_sles_packages }}"
       when: ansible_distribution_file_variety == "SUSE"
+    # Enable iscsid and multipathd service
+    - role: csm.enable_iscsid_multipathd
+
   tasks:
     - name: Enable smart service
       systemd:
@@ -148,3 +151,12 @@
         name: cray-node-exporter
         state: started
         enabled: true
+
+# Compute-nodes only play
+- hosts: Compute:!cfs_image
+  gather_facts: no
+  any_errors_fatal: true
+  remote_user: root
+  roles:
+    # Enable iscsid and multipathd service
+    - role: csm.enable_iscsid_multipathd
```

> Step-5: Commit above changes and push the changes to VCS

Reviewer 3: Change to "Commit **_the_** changes and push **_them_** to VCS"

**Example**

```bash
ncn-m001:/home/csm-config-management # git add csm_packages.yml
ncn-m001:/home/csm-config-management # git commit -m "fix for CASMTRIAGE-7509"
ncn-m001:/home/csm-config-management # git push --set-upstream origin CASMTRIAGE-7509

> ncn-m001:/home/csm-config-management # COMMIT=`git log | head -n 1 | awk '{print $2}'`

Reviewer 3: A simpler command is:
```
COMMIT=$(git log -1 --pretty='format:%H')
```

ncn-m001:/home/csm-config-management # echo $COMMIT
bf214b8a9867531a38f8ca28b6ffae1fe56724ce
```

Step-6: Create new CFS configuration with above change to be applied

**Example**

```bash
ncn-m001:/home/csm-config-management # SESSIONTEMPLATE=compute-25.1.0-alpha2.x86_64-csm-160-rc4
ncn-m001:/home/csm-config-management # CFS_CONFIG=`cray bos sessiontemplates describe $SESSIONTEMPLATE --format json | jq -r .cfs.configuration`
ncn-m001:/home/csm-config-management # cray cfs configurations describe $CFS_CONFIG --format json | jq '. | del(.lastUpdated) | del(.name)'  > $CFS_CONFIG
```

> Step-7: Update commit id ($COMMIT from step 5 in $CFS_CONFIG for “csm_packages.yml”)

Reviewer 3: Put both `$COMMIT` and `$CFS_CONFIG` in back-ticks, and change to back-tics for `csm_packages.yml`.

```bash
ncn-m001:/home/csm-config-management # vim $CFS_CONFIG
```

```bash
ncn-m001:/home/csm-config-management # cat $CFS_CONFIG
{
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "bf214b8a9867531a38f8ca28b6ffae1fe56724ce", 
      "name": "csm-packages-1.6.0-rc.4",
      "playbook": "csm_packages.yml"
    },	
…
```

Step-8: Update cfs config

```bash
ncn-m001:/home/csm-config-management # cray cfs configurations update --file $CFS_CONFIG $CFS_CONFIG
```

Step-9: Create new BOS session template with this new config change

> Please refer: [Create BOS session template for iSCSI SBPS](../boot_orchestration/Create_a_Session_Template_to_Boot_Compute_Nodes_with_SBPS.md)

Reviewer 3: Change to: Please refer **_to_**:

...
Nov 07 10:30:43 nid000004 iscsid[22286]: iscsid: connect to 10.253.0.3:3260 failed (Connection refused)
...
```

## Workaround Description

### Step-1: Get the version of the “csm-packages” from compute node BOS session template

**Example**

Find the session template name:

```bash
ncn-m001:~ # cray bos sessiontemplates  list | grep compute-*
…
enable_cfs = true
name = "compute-25.1.0-alpha2.x86_64-csm-160-rc4”
…
```

Find the `configuration` name from the sessiontemplates describe:

```bash
ncn-m001:~ # cray bos sessiontemplates describe compute-25.1.0-alpha2.x86_64-csm-160-rc4  --format json
{…
  "cfs": {
    "configuration": "compute-25.1.0-alpha2-csm-160-rc4"
  },
…
}
```

Use the `configuration` value to describe the configuration:

```bash
ncn-m001:~ # cray cfs configurations describe compute-25.1.0-alpha2-csm-160-rc4 --format json
{
  "lastUpdated": "2024-11-02T12:15:21Z",
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "d530f0a277c9d5dc9e3cb487d32d6b316757f00e",
      "name": "csm-packages-1.6.0-rc.4", 
      "playbook": "csm_packages.yml"
    }, 
…
```

> The `name` from the describe above identifies product catalog, the version is after `csm-packages-`

Reviewer 4: Is this a better way to word this statement?

```suggestion
The `name` from the describe above identifies the product catalog. Use the version after `csm-packages-` in the next step.
```

# Boot content projection services can fail if `iscsid` and `multipathd` services are not enabled on compute nodes and UANs

## Issue Description

iSCSI based boot content projection which is also known as "Scalable Boot Content Projection" (SBPS) for `rootfs` and `PE` images

> is supported in version CSM 1.6.0 and above. On a customer system, using `CSM-1.6.0` with `USS-1.1.x` on compute nodes/ UANs in order to support AARCH64 images,

Reviewer 5: ```suggestion
is supported in CSM version 1.6.0 and later. On a customer system, using `CSM-1.6.0` with `USS-1.1.x` on compute nodes/ UANs in order to support AARCH64 images,
```

> `iscsid` and `multipathd` services are not enabled by default. SBPS will not be resilient across worker node reboots if these services are not enabled by default on compute/ UANs.

Reviewer 5: ```suggestion
`iscsid` and `multipathd` services are not enabled by default. SBPS will not be resilient across worker node reboots if these services are not enabled by default on compute nodes or UANs.
```

## Issue Identification

This issue can be identified by the following symptoms:

> On a compute/UAN node (iSCSI Initiator) we can observe the following SQUASHFS error messages in the console log:

Reviewer 5: ```suggestion
On a compute node or UAN (iSCSI Initiator) we can observe the following SQUASHFS error messages in the console log:
```
I would reword this. UAN node is redundant.

> ```text
> nid000004:~ # dmesg -T | grep "SQUASHFS error" | head -n  1
> [Sat Nov  2 22:32:41 2024] SQUASHFS error: xz decompression failed, data probably corrupt
> ```
> 

Reviewer: ````suggestion
```bash
dmesg -T | grep "SQUASHFS error" | head -n  1
```

Example output:

```
[Sat Nov  2 22:32:41 2024] SQUASHFS error: xz decompression failed, data probably corrupt
```

````

> On a compute/UAN node (iSCSI Initiator) we can observe that the `iscsid` service is not active:

Reviewer 5: ```suggestion
On a compute node or UAN (iSCSI Initiator) we can observe that the `iscsid` service is not active:
```

# Boot content projection services can fail if `iscsid` and `multipathd` services are not enabled on compute nodes and UANs

## Issue Description

iSCSI based boot content projection which is also known as "Scalable Boot Content Projection" (SBPS) for `rootfs` and `PE` images
is supported in CSM version CSM 1.6.0 and later. On a customer system, using `CSM-1.6.0` with `USS-1.1.x` on compute nodes/ UANs in order to support AARCH64 images,
`iscsid` and `multipathd` services are not enabled by default. SBPS will not be resilient across worker node reboots if these services are not enabled by default on compute nodes or UANs.

## Issue Identification

This issue can be identified by the following symptoms:

On a compute node or UAN (iSCSI Initiator) we can observe the following SQUASHFS error messages in the console log:

```bash
dmesg -T | grep "SQUASHFS error" | head -n  1
```

Example output:

```text
[Sat Nov  2 22:32:41 2024] SQUASHFS error: xz decompression failed, data probably corrupt
```

On a compute node or UAN (iSCSI Initiator) we can observe that the `iscsid` service is not active:

```bash

> ncn-s004# systemctl status iscsid

Reviewer: The prompt is for a storage node but the preceding text says "on a compute or UAN."

Nov 07 10:25:14 nid000004 iscsid[22286]: iscsid: connect to 10.253.0.3:3260 failed (No route to host)
...
Nov 07 10:30:43 nid000004 iscsid[22286]: iscsid: connect to 10.253.0.3:3260 failed (Connection refused)
...
```

## Workaround Description

### 1: Get the version of the `csm-packages` from compute node BOS session template

Example:

Find the session template name:

```bash
ncn-m001:~ # cray bos sessiontemplates  list | grep compute-*
…
enable_cfs = true
name = "compute-25.1.0-alpha2.x86_64-csm-160-rc4"
…
```

Find the `configuration` name from the `sessiontemplates` describe:

```bash
ncn-m001:~ # cray bos sessiontemplates describe compute-25.1.0-alpha2.x86_64-csm-160-rc4  --format json
{…
  "cfs": {
    "configuration": "compute-25.1.0-alpha2-csm-160-rc4"
  },
…
}
```

Use the `configuration` value to describe the configuration:

```bash
ncn-m001:~ # cray cfs configurations describe compute-25.1.0-alpha2-csm-160-rc4 --format json
{
  "lastUpdated": "2024-11-02T12:15:21Z",
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",
      "commit": "d530f0a277c9d5dc9e3cb487d32d6b316757f00e",

>       "name": "csm-packages-1.6.0-rc.4", 

Reviewer: There is an unrecognized character here.

ncn-m001:/home/csm-config-management # git commit -m "fix for CASMTRIAGE-7509"
ncn-m001:/home/csm-config-management # git push --set-upstream origin CASMTRIAGE-7509

ncn-m001:/home/csm-config-management # COMMIT=$(git log -1 --pretty='format:%H')
ncn-m001:/home/csm-config-management # echo $COMMIT
bf214b8a9867531a38f8ca28b6ffae1fe56724ce
```

### 6: Create new CFS configuration with above change to be applied

Example:

```bash
ncn-m001:/home/csm-config-management # SESSIONTEMPLATE=compute-25.1.0-alpha2.x86_64-csm-160-rc4
ncn-m001:/home/csm-config-management # CFS_CONFIG=`cray bos sessiontemplates describe $SESSIONTEMPLATE --format json | jq -r .cfs.configuration`
ncn-m001:/home/csm-config-management # cray cfs configurations describe $CFS_CONFIG --format json | jq '. | del(.lastUpdated) | del(.name)'  > $CFS_CONFIG
```

### 7: Update commit id (`$COMMIT` from `step 5` in `$CFS_CONFIG` for `csm_packages.yml`)

```bash
ncn-m001:/home/csm-config-management # vim $CFS_CONFIG
```

```bash
ncn-m001:/home/csm-config-management # cat $CFS_CONFIG
{
  "layers": [
    {
      "cloneUrl": "https://api-gw-service-nmn.local/vcs/cray/csm-config-management.git",

>       "commit": "bf214b8a9867531a38f8ca28b6ffae1fe56724ce", 

Reviewer: Another malformed unicode character.

      "name": "csm-packages-1.6.0-rc.4",
      "playbook": "csm_packages.yml"
    },
…
```

### 8: Update `cfs` config

```bash
ncn-m001:/home/csm-config-management # cray cfs configurations update --file $CFS_CONFIG $CFS_CONFIG

> ```

Reviewer: Another prompt fix.
